tidalapi/page.py

```diff
@@ -118,25 +177,23 @@ def __init__(self, session: "tidalapi.session.Session"):
             "MIX_LIST": self.session.parse_mix,
         }
 
-    def parse(self, json_obj):
+    def parse(self, json_obj: JsonObj) -> AllCategories:
         result = None
         category_type = json_obj["type"]
         if category_type in ("PAGE_LINKS_CLOUD", "PAGE_LINKS"):
-            category: Union[
-                PageLinks, FeaturedItems, ItemList, TextBlock, LinkList
-            ] = PageLinks(self.session)
+            category: PageCategories = PageLinks(self.session)
         elif category_type in ("FEATURED_PROMOTIONS", "MULTIPLE_TOP_PROMOTIONS"):
             category = FeaturedItems(self.session)
         elif category_type in self.item_types.keys():
             category = ItemList(self.session)
         elif category_type == "MIX_HEADER":
-            return self.session.parse_mix(json_obj["mix"])
+            return cast("Mix", self.session.parse_mix(json_obj["mix"]))
```

Comment:
why do we need the cast?  Can't we just typehint `parse_mix`?

Reply:
This line causes mypy to fall over in strict mode, so I added the explicit casts to quell some of the errors.

> tidalapi/page.py:191: error: Returning Any from function declared to return "Artist | Album | PageLinks | FeaturedItems | ItemList | TextBlock | LinkList | Mix"  [no-any-return]

I think cleaning up `Session` will help improve the downstream inferences.

Reply:
makes sense; leave for now then.  casts are a big code smell so they'll draw attention to it.

Reply:
Actually, looks like you can do it in session.

in the constructor:

```python
        self.parse_mix: Callable[..., mix.Mix] = self.mix().parse
```

And then drop the casts here.

I'd prefer that as it's providing the information where it's needed.  (I'd prefer not dynamically constructing these methods but that can wait unless you'd rather rewrite them as:

```python
def parse_mix(self, obj: dict) -> Mix:
    return self.mix().parse(obj)
```

Which I guess is rather verbose.)

Separately, how are you testing?  `mypy --strict tidalapi/page.py` throws 5 errors for me.

Reply:
> Actually, looks like you can do it in session.

Unfortunately that doesn't work. It should, but mypy is giving up somewhere. Refactoring them into standalone methods worked. Along the way, I discovered an issue with some of the earlier session typings I introduced: I was confused by/following the docstrings and assumed we re-exported the models from the root module. We don't :facepalm: . I reparented the references to point to the relevant model modules.